### PR TITLE
Polarity based direction heuristics

### PR DESCRIPTION
**Pull request by anwu1219**
Add method to compute the polarity of a Relu and bias the fixing and case split order based on it.
Tested on 60 acas benchmarks and 40 MNIST benchmarks. 
Solved 11 more SAT and 3 more UNSAT with a 20 minutes timeout compared to the current master.

### CODE REVIEW

File: src/configuration/GlobalConfiguration.h [hunk 1]
```diff
@@ -21,6 +21,8 @@ class GlobalConfiguration
 public:
     static void print();
 
+    static const bool USE_POLARITY_BASED_DIRECTION_HEURISTICS;
```
Comment (ahmed-irfan) [inline src/configuration/GlobalConfiguration.h]: please add some description

File: src/engine/Engine.cpp [hunk 2]
```diff
@@ -1924,6 +1925,15 @@ void Engine::checkOverallProgress()
     }
 }
 
+void Engine::updateDirections()
+{
+    if ( GlobalConfiguration::USE_POLARITY_BASED_DIRECTION_HEURISTICS )
+        for ( const auto &constraint : _plConstraints )
+            if ( constraint->supportPolarity() &&
+                 constraint->isActive() && !constraint->phaseFixed() )
+                ( (ReluConstraint *) constraint )->updateDirection();
```
Comment (ahmed-irfan) [inline src/engine/Engine.cpp]: here you are making the assumption that if the plconstraint supports polarity then it must be a reluconstraint. right?

File: src/engine/ReluConstraint.cpp [hunk 3]
```diff
@@ -275,14 +276,30 @@ List<PiecewiseLinearConstraint::Fix> ReluConstraint::getPossibleFixes() const
...
+                fixes.append( PiecewiseLinearConstraint::Fix( _b, fValue ) );
+                fixes.append( PiecewiseLinearConstraint::Fix( _f, 0 ) );
+            }
         }
     }
     else
     {
-        fixes.append( PiecewiseLinearConstraint::Fix( _b, 0 ) );
-        fixes.append( PiecewiseLinearConstraint::Fix( _f, bValue ) );
+        if ( _direction == PHASE_ACTIVE )
```
Comment (ahmed-irfan) [inline src/engine/ReluConstraint.cpp]: this and the above code could be merged by rewriting the parent if-else block

File: src/engine/ReluConstraint.cpp [hunk 4]
```diff
@@ -415,6 +432,19 @@ List<PiecewiseLinearCaseSplit> ReluConstraint::getCaseSplits() const
 
     List<PiecewiseLinearCaseSplit> splits;
 
+    if ( _direction == PHASE_INACTIVE )
+    {
+        splits.append( getInactiveSplit() );
+        splits.append( getActiveSplit() );
+        return splits;
+    }
+    else if ( _direction == PHASE_ACTIVE )
```
Comment (ahmed-irfan) [inline src/engine/ReluConstraint.cpp]: remove else

File: src/engine/ReluConstraint.cpp [hunk 5]
```diff
@@ -829,6 +864,21 @@ unsigned ReluConstraint::getAux() const
     return _aux;
 }
 
+double ReluConstraint::computePolarity() const
+{
+    double currentLb = _lowerBounds[_b];
+    double currentUb = _upperBounds[_b];
+    if ( currentLb >= 0 || currentUb <= 0 ) return 1;
```
Comment (ahmed-irfan) [inline src/engine/ReluConstraint.cpp]: why currentUb <= 0?

File: src/engine/ReluConstraint.cpp [hunk 6]
```diff
@@ -829,6 +864,21 @@ unsigned ReluConstraint::getAux() const
...
+    double width = currentUb - currentLb;
+    double sum = currentUb + currentLb;
+    return sum / width;
```
Comment (ahmed-irfan) [inline src/engine/ReluConstraint.cpp]: do we really need to divide the sum by width? isn't sum enough?

**Comment by guykatzz** [comment]
Hi Andrew,
I also reviewed the changes. I think Ahmed raised some good points. In addition:
- When the new polarity flag is off, does the ReLU constraint have the same logic as before?
- We should definitely remove the relu constraint cast in Engine.cpp that Ahmed commented on
- The way polarity is computed is not intuitive; can you perhaps add a short explanation somewhere? Or was this decided empirically?

File: src/engine/ReluConstraint.cpp (hunk 5 above)
Comment (ahmed-irfan) [reply src/engine/ReluConstraint.cpp]: is it `if ( currentLb >= 0 && currentUb <= 0 )` ?

File: src/configuration/GlobalConfiguration.h (hunk 1 above)
Comment (anwu1219) [reply src/configuration/GlobalConfiguration.h]: Added some description.

File: src/engine/Engine.cpp (hunk 2 above)
Comment (anwu1219) [reply src/engine/Engine.cpp]: Added updateDirection as a virtual method in PiecewiseLinearConstraint.h that does nothing by default.

File: src/engine/ReluConstraint.cpp (hunk 5 above)
Comment (anwu1219) [reply src/engine/ReluConstraint.cpp]: Whoops this is a bug. Fixed it and added more explanation in the header.

File: src/engine/ReluConstraint.cpp (hunk 6 above)
Comment (anwu1219) [reply src/engine/ReluConstraint.cpp]: We just need the sign for direction heuristic. However, we need this to compute how symmetric the bound is with respect to zero in the relu-splitting branching heuristic for DnC. Concretely, we will pick ReLUs with the polarity that is closest to 0.

**Comment by anwu1219** [comment]
@guykatzz 
- yes, when that option is off, the ReLU constraint has the same logic as before.
- Addressed.
- I added a description of how polarity is computed in the ReLuConstraint header file.

**Comment by ahmed-irfan** [comment]
@anwu1219 maybe you forgot to push your commits

File: src/engine/ReluConstraint.cpp (hunk 3 above)
Comment (ahmed-irfan) [reply src/engine/ReluConstraint.cpp]: after a discussion with @anwu1219 , it seems what I had in my mind is not possible.

**Comment by anwu1219** [comment]
Solves 11 more SAT and 3 more UNSAT

File: src/engine/PiecewiseLinearConstraint.h [hunk 7]
```diff
@@ -203,6 +203,23 @@ class PiecewiseLinearConstraint : public ITableau::VariableWatcher
...
+      Return true if and only if this piecewise linear constraint supports
+      the polarity metric
+    */
+    virtual bool supportPolarity() const
+    {
+        return false;
+    }
+
+    /*
+      Return true if and only if this piecewise linear constraint supports
```
Comment (guykatzz) [inline src/engine/PiecewiseLinearConstraint.h]: Comment seems wrong

File: src/engine/ReluConstraint.cpp [hunk 8]
```diff
@@ -829,6 +864,22 @@ unsigned ReluConstraint::getAux() const
     return _aux;
 }
 
+double ReluConstraint::computePolarity() const
+{
+    double currentLb = _lowerBounds[_b];
+    double currentUb = _upperBounds[_b];
+    if ( currentLb >= 0 ) return 1;
+    if (  currentUb <= 0 ) return -1;
```
Comment (guykatzz) [inline src/engine/ReluConstraint.cpp]: remove extra space

File: src/engine/ReluConstraint.h [hunk 9]
```diff
@@ -162,12 +162,42 @@ class ReluConstraint : public PiecewiseLinearConstraint
...
+      the bound of the input to this ReLU is with respect to 0.
+      Let LB be the lowerbound, and UB be the upperbound.
+      If LB >= 0, polarity is 1.
+      If UB <= 0, polarity is -1.
+      If LB < 0, and UB > 0, polarity is ( LB + UB ) / (UB - LB).
+
+      We divide the sum by the width of the interval so that the polarity is
+      always between -1 and 1. The closer it is to 0, the more symmetric the
+      bound is.
+
```
Comment (guykatzz) [inline src/engine/ReluConstraint.h]: remove empty line

File: src/engine/ReluConstraint.h [hunk 10]
```diff
@@ -162,12 +162,42 @@ class ReluConstraint : public PiecewiseLinearConstraint
...
+    */
+    double computePolarity() const;
+
+    /*
+      Update the preferring direction for fixing and handling case split
+    */
+    void updateDirection();
+
```
Comment (guykatzz) [inline src/engine/ReluConstraint.h]: remove empty line

File: src/engine/ReluConstraint.h [hunk 11]
```diff
@@ -162,12 +162,42 @@ class ReluConstraint : public PiecewiseLinearConstraint
...
     bool _auxVarInUse;
     unsigned _aux;
 
+    /*
+      Denotes which case split to handle first.
+      And which phase status to repair a relu into.
+    */
+    PhaseStatus _direction;
+
+
```
Comment (guykatzz) [inline src/engine/ReluConstraint.h]: remove empty line

File: src/engine/ReluConstraint.h [hunk 12]
```diff
@@ -162,12 +162,42 @@ class ReluConstraint : public PiecewiseLinearConstraint
...
+      We divide the sum by the width of the interval so that the polarity is
+      always between -1 and 1. The closer it is to 0, the more symmetric the
+      bound is.
+
+    */
+    double computePolarity() const;
+
+    /*
+      Update the preferring direction for fixing and handling case split
```
Comment (guykatzz) [inline src/engine/ReluConstraint.h]: preferring --> preferred?

File: src/engine/PiecewiseLinearConstraint.h (hunk 7 above)
Comment (anwu1219) [reply src/engine/PiecewiseLinearConstraint.h]: Corrected the comment

**Pull request merged**